File: 92. Reverse Linked List II-1.cpp
struct ListNode {
  int val;
  ListNode *next;
  ListNode() : val(0), next(nullptr) {}
  ListNode(int x) : val(x), next(nullptr) {}
  ListNode(int x, ListNode *next) : val(x), next(next) {}
};

class Solution {
public:
  ListNode *reverseBetween(ListNode *head, int left, int right) {
    if (left == right)
      return head;

    ListNode dummy;
    dummy.next = head;

    ListNode *front = &dummy;
    for (int i = 0; i < left - 1; ++i)
      front = front->next;

    ListNode *end = &dummy;
    for (int i = 0; i <= right; ++i)
      end = end->next;

    ListNode *mid = front->next;
    ListNode *prev = nullptr;
    ListNode *cur = mid;
    while (cur != end) {
      ListNode *nxt = cur->next;
      cur->next = prev;
      prev = cur;
      cur = nxt;
    }

    front->next = prev;
    mid->next = end;

    return dummy.next;
  }
};

File: 92. Reverse Linked List II-1.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def reverseBetween(
        self, head: Optional[ListNode], left: int, right: int
    ) -> Optional[ListNode]:
        dummy = ListNode(next=head)
        front = dummy
        for _ in range(left - 1):
            front = front.next

        prev = None
        cur = front.next
        for _ in range(right - left + 1):
            nxt = cur.next
            cur.next = prev
            prev = cur
            cur = nxt

        front.next.next = cur
        front.next = prev

        return dummy.next
